src/js_realm.hpp

```diff
@@ -1382,41 +1382,151 @@ void RealmClass<T>::compact(ContextType ctx, ObjectType this_object, Arguments&
     return_value.set(realm->compact());
 }
 
+/**
+ * @brief Create a copy of one realm file to another realm file.
+ *  Conversion between synced and non-synced realms is supported.
+ *  Invocation of `writeCopyTo` is overloaded to two scenarios:
+ *  1) `writeCopyTo(<path: string>, [encryption key: string])`, which supports copying a local realm
+ *    to another local realm, or converting a synced realm to a local realm.
+ *  2) `writeCopyTo(<config: object>)`, which, in addition to the above, supports conversion of a local
+ *    realm to a synced realm if the `sync` section is present in `config`.
+ *
+ * @param ctx JS enviroment context
+ * @param this_object Realm object passed from JS
+ * @param args Either `<string>, [string]`, or `object` -- see function brief.
+ * @param return_value none
+ */
 template <typename T>
 void RealmClass<T>::writeCopyTo(ContextType ctx, ObjectType this_object, Arguments& args, ReturnValue& return_value)
 {
     args.validate_maximum(2);
 
+    enum CopyMode {
```

Comment:
It's probably more of a reflection on the kind of software I've written (I've not come across that much "working with bits" code, so I'm not that comfortable reading it), and I am quite possibly missing something quite obvious, but I'd be interested to know the advantages of doing it this way versus something more "explicit" e.g.

```
enum CopyType {
  WithPath,
  WithConfig
};

enum CopyDestination {
  ToSync,
  ToLocal
};

CopyType copy_type = CopyType.WithPath;
CopyDestination copy_destination = CopyDestination.ToSync; 
bool copy_encrypted = false;

// ...

if (Value::is_object(ctx, firstParamValue)) {
  copy_type = CopyType.WithConfig;
}

// ... 

switch (copy_type) {
  case CopyType.WithConfig:
    // ...
    if (Value::is_valid(syncConfigValue) && Value::is_object(ctx, syncConfigValue)) {
      copy_destination = CopyDestination.ToSync;
    }
    else {
     copy_destination = CopyDestination.ToLocal;
    }
    break;
  case CopyType.WithPath:
    // ....
}

// etc ...

if (copy_destination == CopyDestination.ToLocal) {
  realm->write_copy(output_path, encryption_key.get());
}
else if (copy_destination == CopyDestination.ToSync) {
  // ... 
}
````

Reply:
N/A in the refactored implementation